content/en/blogs/devops-automatic-testing.md

```diff
@@ -0,0 +1,247 @@
+
+The built-in DevOps system of KubeSphere smoothly combines application development and automated release with the container platform. It also supports the integration of third-party private image registries and code repositories, which brings an improved CI/CD under private scenarios and provides the end-to-end user experience.
+
+Nevertheless, it is rarely known to users that KubeSphere DevOps can also be used to build automated testing systems, so that the automated **unit testing, API testing, and UI testing** can have great convenience and the work of testers can be more efficient.
+
+## Unit Testing
+
+Unit testing has a high running frequency that it should be triggered with every code commit. Due to few dependenies, unit testing usually only requires one container runtime environment.
+
+The following is an example of running a unit testing in a pipeline using golang:latest.
+
+```
+pipeline {
+  agent {
+    node {
+      label 'go'
+    }
+  }
+  stages {
+    stage('testing') {
+      steps {
+        container('go') {
+          sh '''
+          git clone https://github.com/etcd-io/etcd.git
+          cd etcd
+          make test
+          '''
+        }
+
+      }
+    }
+  }
+}
+```
+
+You can see the pipeline logs as below:
+
+![golang-test](/images/blogs/en/devops-automatic-testing/golang-test.png)
+
+As for other languages and frameworks, unit testing can also be easily run on Kubernetes by installing certain packages and Mock-related services. The techniques of writing unit testing should be digged deeper rather than runtimes or unit testing plans.
+
+## API Testing
+
+If a team is just getting started with its automated testing, automated API testing would be a great starting point. 
+
+The unit testing is mainly written by R&D folks. In a rapid iteration process, the experienced R&D folks will not forget to write unit testing. Testing plays a much more important role in the faster process of refactoring and making changes. The unit testing is overlooked if no one cares to write any, and it is quite difficult to propel something that is neglected by the practitioners and has limited benefits to show to the managerial folks.
+
+Besides, automated UI testing is usually replaced by manual testing. Meanwhile, the maintenance of automated UI testing brings high costs, which means the automated UI testing in the process of rapid iteration should be performed reasonably.
+
+The advantage of API testing under the architecture of front-end and back-end seperation lies in the relatively complete API-related documents and materials and the team members' relative familiarity with the API, which will be helpful for the testing. 
```

Comment:
->
API tests are more preferable if team members are more familiar with APIs as they run tests with complete API documents and materials as reference, especially in an architecture with separate frontend and backend code.